### ISSUE TITLE
Warn against using unpack operator within components

### ISSUE THREAD

**Issue by Archmonger** [idom-team/flake8-idom-hooks#11]
IDOM core currently has no way of providing "key does not exist for dynamic children" warnings when using the unpack operator.

We'll need to handle this within flake8 by following the AST and identifying any unpack operations that may happen within a component syntax tree.

See here for the original discussion: https://github.com/idom-team/idom/discussions/607

**Comment by rmorshea** [comment]
The challenge here is in figuring out when to warn. Not all usages of `*args` in a component are invalid. For example,

```python
@idom.component
def Example():
    do_something(*args)  # ok
    return html.div(*children)  # no ok
```

One could use the heuristic that anything in the `return` statement must avoid unpacking but that has further complications:

```python
@idom.component
def Example():
    an_element = html.div(*some_children)  # not ok
    return html.div(
        an_element,
        make_an_element(*args)  # ok if `args` is not a list of children
    )
```

My worry here is that one of the following things happens:

1. We don't warn enough, in which case the user gets a false sense of security by turning this check on.
2. We warn too much, in which case users may just turn off this check entirely.

I've been tempted to use the `div[...]` syntax because `div[*args]` is invalid syntax. Thus, it's impossible to do the wrong thing.

**Comment by Archmonger** [comment]
But isn't `div[*(args)]` valid syntax? I utilize that all over the Django parts of Conreq. So in my case I would have never received a warning.

**Comment by rmorshea** [comment]
I get a syntax error:
```python
>>> div[*(args)]
  File "<stdin>", line 1
    div[*(args)]
               ^
SyntaxError: invalid syntax
```

**Comment by Archmonger** [comment]
You're right. It's only during the initialization of an array where [I was using the unpack operator](https://github.com/Archmonger/Conreq/blob/6d8f302cb88f30f6d835371d6f766e412ac3edea/conreq/settings.py#L218).

Wouldn't the syntax you're suggesting make constructing elements awkward? I'm envisioning something like....

```py
div[0] = Element1
div[1] = Element2
```

Is it even possible to use the index operator to denote a list of elements like the previous syntax?

```py
div[ Element1, Element2 ]
```

**Comment by rmorshea** [comment]
The following seems to work:

```python
>>> class X:
...     def __getitem__(self, item):
...         return item
... 
>>> x = X()
>>> x[1, 2, 3, 4, 5, 6, 7, 8, 9]
(1, 2, 3, 4, 5, 6, 7, 8, 9)
```

**Comment by Archmonger** [comment]
That's a pretty clever way to solve this one, I say lets do it.

Perhaps v1->v2 target? Unless it's a super quick implementation.

**Comment by rmorshea** [comment]
Quick to implement if a deprecation warning is included until the old syntax is dropped.

I worry about the syntax being off-putting though. React also has a similar issue with the Javascript spread operator if you're not using JSX. Though, everyone uses JSX so it doesn't really come up.

**Comment by rmorshea** [comment]
I guess the whole framework is probably odd to Python devs though, so maybe this isn't so bad.

**Comment by Archmonger** [comment]
> I guess the whole framework is probably odd to Python devs though, so maybe this isn't so bad.

Agreed. IDOM is pretty far removed from Python styling so it doesn't really matter how much you continue to deviate from that at this point. I generally take the stance that frameworks should protect developers from ever writing bad code (when possible).

Visually, the styling doesn't look "ugly"

```python
# Old
def homepage(websocket):
    state, set_state = idom.hooks.use_state(HomepageState())

    return div(
        navbar(websocket, state, set_state),
        modal(websocket, state, set_state),
        sidebar(websocket, state, set_state),
        viewport_loading(websocket, state, set_state),
        viewport_primary(websocket, state, set_state),
        viewport_secondary(websocket, state, set_state),
    )
```
```python
# New
def homepage(websocket):
    state, set_state = idom.hooks.use_state(HomepageState())

    return div[
        navbar(websocket, state, set_state),
        modal(websocket, state, set_state),
        sidebar(websocket, state, set_state),
        viewport_loading(websocket, state, set_state),
        viewport_primary(websocket, state, set_state),
        viewport_secondary(websocket, state, set_state),
    ]
```
